43 alternate extent field mapping

alexdryden opened the pull request:
Opening pr against the base UIUC specific changes branch to better facility discussing it--opening against main includes all of the UIUC changes and makes it more difficult to read.

alexdryden committed:
test: add tests for parsing alternative extent statements

alexdryden committed:
fix: add option parameter to specify what default value to use

alexdryden committed:
feat: add method to get all of the ASpace extent types

This allows us to iterat through all of the existing extent types
so we can find the best match when parsing the alternative extent
statement to generate structured extents

alexdryden committed:
test: add tests for the mapping method

alexdryden committed:
feat: add methods to support parsing structured exted from string

This will support additional logic in the addResourceExtent method
to parse the alternative extent statments in cases where they were
used in Archon to document multiple extents using unstructured
language. Archon does not permit multiple extents, so the
alternative extent field was widely used to doument multiple extent
types.

The current behavior in addResourceExtent tread the alternative
extent closer to how it is defined in
https://saa-ts-dacs.github.io/dacs/06_part_I/03_chapter_02/05_extent.html
as an alternative way to describe the whole extent.

alexdryden committed:
feat: add stub for mapping extent types

alexdryden committed:
fix: portion should default to "part" if an alt extent exists

In practice, alt extent is used to descibe collections with
multiple extents. The presense of an alt extent should, in most
cases, be an indicator that the extents only describe part of the
collection. The exception is when the alternative extent is
obviously a subsidiary of the main extent, e.g., folders and items
where items count the number of things found in the folders.

graykr commented:
Hi Alex, This all looks like a reasonable start to me -- though, I think we may need to account for string that might be something like "6 microfilm reels and an oversize folder" (no number, but the word "a" or "an" indicating "1").  Or we can look for those in Archon and just correct them to be a number prior to migration. But that can also come later as a potential small feature.

alexdryden commented:
Sounds good! I think I had something to that extend drafted. I'm out Friday, so I'll work on this today, and hopefully wrap it up.

alexdryden committed:
test: add test coverage for using written text for unit values

alexdryden committed:
feat: detect indefinate articles and written numbers as unit values

alexdryden committed:
fix: detect unit values at the start of the string

This prevents catching strings that are just descriptions but
include a number somewhere, like the test case :
"including photostats of 1934 book", which would return 1934 as
unit value and book as unit.

alexdryden committed:
feat: populate the reverse lookup for value enums

alexdryden committed:
feat: add method for reverse enum lookup

alexdryden committed:
feat: add method to get all of the archon extent types

alexdryden committed:
feat: do basic matching for parsed extents against all extents

alexdryden committed:
fix: use string matching method instead of ==

alexdryden committed:
doc: update javadocs

alexdryden committed:
fix: update tests with examples for simple matching that will pass

alexdryden committed:
fix: check if the target extent appears anywhere in the input

alexdryden committed:
test: add tests for processAlternativeExtent

alexdryden committed:
test: update tests to reflect getting a numeral for written values

alexdryden committed:
feat: fixes #43 by adding method to parse alternative extent

into individual strucutred extent statements.

alexdryden committed:
fix: move structured extent and error reporting into if/else block

alexdryden committed:
feat: add error reporting for extents that don't map cleanly

alexdryden committed:
fix: correct typo

alexdryden committed:
fix: fix typos

alexdryden commented:
@graykr everything covered by the original scope should be here. I haven't implemented the fuzzy matching yet, and we might want to move the final construction of the structured extent objects out of the `addResourceExtent` method so that it is easier to test/debug.

graykr commented on src/org/nyu/edu/dlts/utils/ASpaceMapper.java:
```diff
@@ -1165,15 +1246,38 @@ private void addResourceExtent(JSONObject record, JSONObject json) throws Except
+                    extentJA.put(extentJS);
+                } else {
+                    String collectionIdentifier = record.getString("CollectionIdentifier");
+                    String debugMessage = "Collection: " + collectionIdentifier + "has an alternative extent statement that didn't map cleanly\n"
```
If available, I add the title as well and/or the Archon ID to make it easier to determine which collection it is

alexdryden replied:
I added the archon id in [5751e04](https://github.com/UIUCLibrary/archon-migration/pull/59/commits/5751e04d03504a9a3d3f6bcd596128000a20814a)

graykr reviewed (commented):
This all looks reasonable to me. With the fuzzy matching process you have planned to implement, I am wonder which would be better -- testing also looking up common plural/non-plural forms (add an 's' or take off an 's' as applicable and try looking for an extent again), or using fuzzy matches.

alexdryden commented:
Oh that's a really good point--I think it might be a better overall solution for now to compile the plural forms of all the known extents.

When I was looking at the fuzzy matching, without using something really really super sophisticated, like a LLM, I think the selection of the most effective algorithm is a tough problem and really depends on the type of errors we are expecting and the kind of "fuzzy" we have in mind. I realized I was getting bogged down reading about various strategies and it may not pay the best dividends. 

I'll use the credential you supplied in Box to get the extents from the other instances and check for anything that has a non-standard plural form, then add that feature in.

alexdryden committed:
feat: add archon id to the error output

alexdryden committed:
fix: use consistent key names

alexdryden committed:
fix(refactor): fix adding/modifying wrong object and refactor

names to reduce confusion

alexdryden committed:
feat: record if the match was exact or not

This is the next best option to a fuzzy match. For non-exact mat-
ches we should include the part of the alt extent statement we are
parsing in the notes for the extent.

alexdryden committed:
fix: create the new extent object inside the for loop

alexdryden committed:
fix: move adding of container_summary inside the for loop

alexdryden committed:
refactor: simplify the cohersion of unit value into digit form

alexdryden committed:
fix: use mapExtentType parse the unit portion of the extent

somehow I totally negleged to incorporate this.

alexdryden committed:
fix: simplify and eliminate wrapper for structured alt extent(s)

alexdryden committed:
fix(cleanup): fix flow logic and general housekeeping

alexdryden committed:
fix: clean both the the input and the extent units for comparison

alexdryden committed:
fix: use the cleaned version of the extent

alexdryden committed:
feat: match the extent even if pluralization isn't the same

We should have a convention about which to use, but for now we
shouldn't be pedantic about not matching item <=> items.

alexdryden committed:
feat: always use the longest match for "contains" comparisons.

We want to avoid preferring one kind of match simply because it is
later in the list of existing extents. For, example when matching
microfilm reels against this list using a contains <|> contains
strategy, we don't want to end up with "reel" because it was the
last to match:

item
microfilm_reel
megabyte
reel

And for the inverse of this list, we don't want to stop mattching
when we get to reel because there might be a better match later on.

graykr commented on src/org/nyu/edu/dlts/utils/ASpaceEnumUtil.java:
```diff
@@ -355,6 +367,26 @@ public String getASpaceExtentType(String atValue) {
+
+       ArrayList<String> archonExtentStrings = new ArrayList<>();
+        for (String string : lookupListValuesToCodes.keySet()) {
+            if (string.startsWith("extent_type_"))
```
It is actually the value that will start with "extent_type_" in lookupListValuesToCodes, not the key. So the if statement should be:
`if (lookupListValuesToCodes.get(string).startsWith("extent_type_"))`

graykr commented on src/org/nyu/edu/dlts/utils/ASpaceMapper.java:
```diff
@@ -1045,20 +1053,192 @@ public JSONObject convertCollection(JSONObject record, HashMap<String, String> c
+     */
+    public JSONObject mapExtentType(String inputExtent) throws JSONException {
+        String cleanInput = inputExtent.toLowerCase().replace("_", " ");
+        //all of the archon extents should have already been added to ASpace. Keeping this 
```
In testing yesterday with ASpace, I did not find this to be the case -- aSpaceExtents was only including those extents hardcoded in ASpaceEnumUtil. When I uncommented this and changed the variable in the for loop that follows to allExtents, it worked much better.

alexdryden replied:
Awesome, thank you!!

alexdryden replied:
@graykr And just a sanity check for me--you found that to work in combination with the fix to the key lookup (`if (string.startsWith("extent_type_"))` into => `if (lookupListValuesToCodes.get(string).startsWith("extent_type_"))`), mentioned [here](https://github.com/UIUCLibrary/archon-migration/pull/59#discussion_r2068751324) right?

graykr replied:
Yes, I was able to get it to work (aside from the folder/folders issue I noted elsewhere). I could try cherry-picking the commit from my test branch to this branch if that would help -- let me know.

graykr commented on src/org/nyu/edu/dlts/utils/ASpaceMapper.java:
```diff
@@ -1045,20 +1053,192 @@ public JSONObject convertCollection(JSONObject record, HashMap<String, String> c
+            //this should catch most common pluralizations, e.g. one has s/es and the other doesn't
+            else if (cleanInput.contains(cleanExtent) || cleanExtent.contains(cleanInput)) {
+                //prefer the longest match, e.g. don't match "microfilm_reels" to "reel" just because "reel" comes after "microfilm_reel"
+                mapping = mapping.length() < cleanExtent.length() ? extent : mapping;
```
In testing with ASpace, this was leading to the singular term being skipped even when it is an exact match, and then the singular form being copied to the container summary, which looks a bit odd:
![image](https://github.com/user-attachments/assets/c81ba5d5-c57c-4027-8b17-4e9d4d7892ca)

alexdryden replied:
Awesome! Thanks you!!

alexdryden replied:
Hmm. I would have thought the exact match would have been caught by the if block? Maybe there was an issue with leading/trailing white space that prevented the exact match?

alexdryden replied:
Oh, no we would need to bail after we match on the if block. Doh!

graykr replied:
So, this is very odd, but I am not finding "folder" in the result for getAllArchonExtents (even though all the other 9 extents in the IHLC archon version are there), so that may also be part of the issue here...

The code you just pushed is working for "1 volume, 16 items" and similar extents.

alexdryden replied:
Very weird! If that is the only one that is missing, I wonder if it is a off-by-one error somewhere. I'll take a gander when I get back from lunch

graykr replied:
You probably saw this, but just in case -- I figured out what was causing this issue and added a note to the ASpaceEnumUtil file about it.

alexdryden commented:
I'm going to ~~add the work I was in the middle of yesterday on setting up error handling for no match anywhere and then~~ <=(skipping that to avoid confusion) push up fixes based on your review. Thank you so much @graykr !!

alexdryden committed:
feat: include logic for various edge cases

alexdryden committed:
fix: don't keep looking if we found an exact match

alexdryden committed:
fix: iterate over values, not keys

alexdryden committed:
fix: aspace extents isn't ever updated, combine them here.

This should be refactored to happen on startup, otherwise we are
re-generating this list each time we map an extent

graykr commented on src/org/nyu/edu/dlts/utils/ASpaceEnumUtil.java:
```diff
@@ -355,6 +367,26 @@ public String getASpaceExtentType(String atValue) {
+    public ArrayList<String> getAllArchonExtents() {
+
+       ArrayList<String> archonExtentStrings = new ArrayList<>();
+        for (String string : lookupListValuesToCodes.keySet()) {
```
So after some more testing, it looks like we actually need to use enumListIDsToValues here (which will be the reverse -- check each key for "extent_type_" and then add the value to the ArrayList to be returned. 

I think what is happening where the match for "folder" isn't working despite it being an extent in Archon is that lookupListValuesToCodes doesn't hold all the extent ids as values -- since some of the terms/keys are the same as those used for the physical container types used in creating finding aids (so "box", "folder", etc.), the extent ids are getting overwritten once those enums are added. So we'd need to use enumListIDsToValues instead since the keys there will all be unique.

alexdryden replied:
Ack, yes you are exactly right. I thought there was some reason not to do it that way from the start, but now can't seem to think of what it might have been.

graykr commented on src/org/nyu/edu/dlts/utils/ASpaceMapper.java:
```diff
@@ -1067,16 +1248,40 @@ private void addResourceExtent(JSONObject record, JSONObject json) throws Except
-            extentJA.put(extentJS);
+                JSONObject structuredExtent = structuredExtents.getJSONObject(i);
+                if ( ! structuredExtent.getBoolean("exactMatch")) {
+                    altExtentJS.put("container_summary", altExtent);    
```
I think it might be better to put the unitValue here. I have an example from RBML where there were two non-exact matches, and then the full alternate extent was added twice, which is probably not what we want:
![image](https://github.com/user-attachments/assets/8993ffbc-480b-4ecf-a260-1371524a910a)
(This is a collection about bookbinding, in case you were wondering about the unusual extent descriptions...)

alexdryden replied:
This is a really useful example. 
The strategy I was working on was similar to what you suggested. Here is what I was thinking for the following cases:
1. If none of the alt extent can be parsed at all, don't add an additional extent and instead put the alt-extent as a container_summary for the main extent. It can't be parsed at all if the entire alt extent:
    -it doesn't split into multiple AND
    -it doesn't have "unit + text" structure OR
    -none of the split statements can be parsed into "unit + text"
2. For ones that match only through an "includes" match, put the number and unit in the container summary, e.g., record "0.5 linear feet with additional oversize animal skins" in the container summary. Though, if you'd rather it leave off the 0.5 we can do that too. 
3. Do the same thing as 2 for unmapped, so "2 other_unmapped" with "2 typecases" as the summary. But again, we can go with just "typecases" if that is what you think is best.
4. If one of many extents in an alt extent doesn't parse into "unit + test", e.g., the middle one in "1 folder, camera parts, 3 reels", still give it an extent but flag extra special it in the errors and record it as "0 other_unmapped" with the entire statement in the container summary.

alexdryden committed:
fix: use enumListIDsToValues for archon extents to avoid collisions

alexdryden committed:
refactor: remove unused code

graykr commented on src/org/nyu/edu/dlts/utils/ASpaceMapper.java:
```diff
@@ -1067,16 +1248,40 @@ private void addResourceExtent(JSONObject record, JSONObject json) throws Except
+                    altExtentJS.put("number", structuredExtent.getString("unitValue"));
+                } else {
+                    altExtentJS.put("extent_type", ASpaceEnumUtil.UNMAPPED);
+                    altExtentJS.put("number", structuredExtent.getString("unitValue"));
```
It looks like ASpace will not copy the collection record if the number within the extent is an empty string. So here there needs to be a check for whether the unitValue is empty and then add the number as "0" in this case.

alexdryden replied:
Ah, good to know. I think this should be covered now by the changes I just pushed up.

alexdryden committed:
fix: fix flow logic to make sure there is a valuewq

alexdryden committed:
fix: restor accidentally deleted brace

graykr commented on src/org/nyu/edu/dlts/utils/ASpaceMapper.java:
```diff
@@ -1045,41 +1053,267 @@ public JSONObject convertCollection(JSONObject record, HashMap<String, String> c
+                    aspaceCopyUtil.addErrorMessage(message); 
+                }
+            }
+        }
```
there needs to be an "else" here to supply the portion of mainextent (whole) when there is no altextent (the collection records are failing to save without the portion indicated for the extent)

graykr commented on src/org/nyu/edu/dlts/utils/ASpaceMapper.java:
```diff
@@ -1045,41 +1053,267 @@ public JSONObject convertCollection(JSONObject record, HashMap<String, String> c
+     
+        JSONObject structuredExtent = new JSONObject();
+
+        String regex = "(^\\d+|\\d+\\.\\d+|a|an|a single|one|two|three|four|five|six|seven|eight|nine|ten)\\s([A-z\s]+)";
```
is it possible to also add regex for decimals without a leading 0, e.g., ".15 megabytes"?

alexdryden committed:
fix: minor issues

graykr commented:
Looks good in my testing with Archon data! I have added the "ready to test with aspace" label to this issue for it to be merged to the test branch this week.  Thanks for all your work on this, Alex!

alexdryden committed:
fix: include only imports that are used in code

alexdryden committed:
chore: add testing library

alexdryden committed:
fix: remove local testing

graykr commented on src/org/nyu/edu/dlts/utils/ASpaceMapper.java:
```diff
@@ -1045,41 +1048,269 @@ public JSONObject convertCollection(JSONObject record, HashMap<String, String> c
         }
 
-        json.put("extents", extentJA);
+        allExtents.put(mainExtent);
```
@alexdryden Do you think it would work to put the main extent into the allExtents variable before adding any of the altExtents, and then rely on it being a reference to the object that might be changed later on? 

Or if there is another way that would be better, that also works -- I was just thinking that this might be the easiest code-wise, assuming that the mainExtent JSONObject in the allExtents JSONArray can be changed after it is added.

Right now we are getting the parsed altExtent being listed first in ASpace and the main extent being listed last. I'm not sure why I didn't notice this earlier, but it really stands out in ArcLight, I guess!

alexdryden replied:
Yeah! Sounds good. Because this has already been merged, to help keep things clean I'll open up a new pr with this issue. To double-check, our branching strategy is to pull down the most recent `test-auto-merged` and make the feature branch from that, right?

graykr replied:
Yes, I think that would make things go more smoothly at this point.

alexdryden replied:
Okay, this should be fixed in https://github.com/UIUCLibrary/archon-migration/pull/72. Some of the parameters for the main extent depend on the alt extent processing (like if it should be whole or part). The simplest solution seemed to be setting the index for where to put the main extent.
